include/oneapi/dpl/experimental/kt/two_pass_scan.h

```diff
@@ -56,45 +57,108 @@ sub_group_masked_scan(const SubGroup& sub_group, MaskOp mask_fn, InitBroadcastId
             value = binary_op(partial_carry_in, value);
         }
     }
-    value = binary_op(init, value);
-    // Adjust for an exclusive scan if requested
-    if constexpr (!Inclusive)
+    if constexpr (InitPresent)
+    {
+        value = binary_op(init_and_carry.__v, value);
+    }
+    LazyValueType old_init;
+    if constexpr (InitPresent)
+    {
+        if (sub_group_local_id == 0)
+            old_init.__setup(init_and_carry.__v);
+        init_and_carry.__v = sycl::group_broadcast(sub_group, value, init_broadcast_id);
+    }
+    else
+    {
+        init_and_carry.__setup(sycl::group_broadcast(sub_group, value, init_broadcast_id));
+    }
+    
+    value = sycl::shift_group_right(sub_group, value, 1);
+    if constexpr (InitPresent)
     {
-        auto old_init = init;
-        init = sycl::group_broadcast(sub_group, value, init_broadcast_id);
-        value = sycl::shift_group_right(sub_group, value, 1);
         if (sub_group_local_id == 0)
-            value = old_init;
+        {
+            value = old_init.__v;
+            old_init.__destroy();
+        }
+    }
+    //return by reference value and init_and_carry
+}
+
+
+template <std::uint8_t VL, bool InitPresent, typename MaskOp, typename InitBroadcastId, typename SubGroup,
+          typename BinaryOp, typename ValueType, typename LazyValueType>
+void
+inclusive_sub_group_masked_scan(const SubGroup& sub_group, MaskOp mask_fn, InitBroadcastId init_broadcast_id, ValueType& value,
+                      BinaryOp binary_op, LazyValueType& init_and_carry)
+{
+
+    std::uint8_t sub_group_local_id = sub_group.get_local_linear_id();
+    _ONEDPL_PRAGMA_UNROLL
+    for (std::uint8_t shift = 1; shift <= VL / 2; shift <<= 1)
+    {
+        auto partial_carry_in = sycl::shift_group_right(sub_group, value, shift);
+        if (mask_fn(sub_group_local_id, shift))
+        {
+            value = binary_op(partial_carry_in, value);
+        }
+    }
+    if constexpr (InitPresent)
```

Comment:
Similar comment to the exclusive scan case. We should be able to combine the two conditionals.

Reply:
Yes, thanks